src/libraries/System.Private.CoreLib/src/System/Reflection/FieldAccessor.cs

```diff
@@ -0,0 +1,512 @@
+                            ref Unsafe.AsRef<short>(_addressOrOffset.ToPointer()),
+                            Unsafe.As<byte, short>(ref value!.GetRawData()));
+                        return;
+
+                    case FieldAccessorType.StaticValueTypeSize4:
+                        VerifyStaticField(ref value, invokeAttr, binder, culture);
+                        Volatile.Write(
+                            ref Unsafe.AsRef<int>(_addressOrOffset.ToPointer()),
+                            Unsafe.As<byte, int>(ref value!.GetRawData()));
+                        return;
+
+                    case FieldAccessorType.StaticValueTypeSize8:
+                        VerifyStaticField(ref value, invokeAttr, binder, culture);
+                        Volatile.Write(
+                            ref Unsafe.AsRef<long>(_addressOrOffset.ToPointer()),
+                            Unsafe.As<byte, long>(ref value!.GetRawData()));
+                        return;
+
+                    case FieldAccessorType.SlowPathUntilClassInitialized:
+                        if (IsStatic())
+                        {
+                            VerifyStaticField(ref value, invokeAttr, binder, culture);
+                        }
+                        else
+                        {
+                            VerifyInstanceField(obj, ref value, invokeAttr, binder, culture);
+                        }
+
+                        isClassInitialized = false;
+                        RuntimeFieldHandle.SetValue(_fieldInfo, obj, value, (RuntimeType)_fieldInfo.FieldType, _fieldInfo.m_declaringType, ref isClassInitialized);
+                        if (isClassInitialized)
+                        {
+                            Initialize();
+                        }
+
+                        return;
+
+                    case FieldAccessorType.NoInvoke:
+                        if (_fieldInfo.DeclaringType is not null && _fieldInfo.DeclaringType.ContainsGenericParameters)
+                            throw new InvalidOperationException(SR.Arg_UnboundGenField);
+
+                        throw new FieldAccessException();
+                }
+            }
+
+            // All other cases use the slow path.
+            if (IsStatic())
+            {
+                VerifyStaticField(ref value, invokeAttr, binder, culture);
+            }
+            else
+            {
+                VerifyInstanceField(obj, ref value, invokeAttr, binder, culture);
+            }
+
+            isClassInitialized = true;
+            RuntimeFieldHandle.SetValue(_fieldInfo, obj, value, (RuntimeType)_fieldInfo.FieldType, _fieldInfo.m_declaringType, ref isClassInitialized);
+        }
+
+        private bool IsStatic() => (_fieldInfo.Attributes & FieldAttributes.Static) == FieldAttributes.Static;
+
+        private void VerifyStaticField(ref object? value, BindingFlags invokeAttr, Binder? binder, CultureInfo? culture)
+        {
+            VerifyInitOnly();
```

Comment:
The InitOnly verification should be only needed for the "until intialized" case. We should not need in the steady state if we flip to steady state only for fields that are not InitOnly.

Reply:
Hmm, it would regress the getter for static InitOnly fields.

Reply:
Yeah I think the getter perf for initonly fields is important. It wouldn't be expected that init-only field access could be much slower.

To avoid the call, we could add additional states for initonly vs. non-initonly, however that doesn't seem worth it to save the simple `if` check.